hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -0,0 +1,403 @@
+      case MakeStream(_, t) => PType.canonical(t)
+      case MakeNDArray(data, shape, _) => {
+        data.inferSetPType(env)
+        shape.inferSetPType(env)
+
+        // TODO: Is this wrong?
+        val nElem = shape.pType2.asInstanceOf[PTuple].size
+
+        // TODO: requiredeness?
+        PNDArray(coerce[PArray](data.pType2).elementType, nElem)
+      }
+      case _: ArrayLen => PInt32()
+      case _: ArrayRange => PArray(PInt32())
+      case _: StreamRange => PStream(PInt32())
+      case _: LowerBoundOnOrderedCollection => PInt32()
+      case _: ArrayFor => PVoid
+      case _: InitOp => PVoid
+      case _: SeqOp => PVoid
+      case _: Begin => PVoid
+      // FIXME in IR suggests Die should give type any, fix this once that is fixed
+      // but in test, TFloat64() is used, which will fit
+      case Die(_, t) => PType.canonical(t)
+      case If(cond, cnsq, altr) => {
+        cond.inferSetPType(env)
+        cnsq.inferSetPType(env)
+        altr.inferSetPType(env)
+
+        assert(cond.pType2 == PBoolean())
+
+        if (cnsq.pType2 != altr.pType2)
+          cnsq.pType2.deepOptional()
+        else {
+          assert(cnsq.pType2 isOfType altr.pType2)
+          cnsq.pType2
+        }
+      }
+      case Let(name, value, body) => {
+        value.inferSetPType(env)
+        body.inferSetPType(env.bind(name, value.pType2))
+        body.pType2
+      }
+      // TODO: It feels like this PType could benefit from inspecting the values it applies the op to
+      // whenever the values are known at time of compilation (current only time of inference)
+      case ApplyBinaryPrimOp(op, l, r) => {
+          l.inferSetPType(env)
+          r.inferSetPType(env)
+
+          val required = l.pType2.required && r.pType2.required
+          val vType = BinaryOp.getReturnType(op, l.pType2.virtualType, r.pType2.virtualType).setRequired(required)
+
+          PType.canonical(vType)
+      }
+      case ApplyUnaryPrimOp(op, v) => {
+        v.inferSetPType(env)
+        println("IN")
+        println(v.pType2)
+        println(v.pType2.required)
+        PType.canonical(UnaryOp.getReturnType(op, v.pType2.virtualType).setRequired(v.pType2.required))
+      }
+      case ApplyComparisonOp(op, l, r) => {
+        l.inferSetPType(env)
+        r.inferSetPType(env)
+        // TODO: should this be l.pType2 == r.pType2 or isOfType
+        assert(l.pType2 isOfType r.pType2)
+        op match {
+          case _: Compare => PInt32(l.pType2.required && r.pType2.required)
+          case _ => PBoolean(l.pType2.required && r.pType2.required)
+        }
+      }
+      case a: ApplyIR => {
+        a.explicitNode.inferSetPType(env)
+        a.pType2
+      }
+      // TODO: this seems wrong:
+      // 1) it may have subtle bugs, .implementation calls ir.typ, which is not pre-filled
+      // 2) .subst sets requiredness on virtualTypes
+      // 3) pType inference isn't really used to any effect here
+//      case a: AbstractApplyNode[_] => {
```

Comment:
rebase -- I added functionality here that will help:
`a.implementation.returnPType(args.map(_.ptype2))`